### PR TITLE
[PATCH v3] linux-gen: cls: fix cos hash-queue handling

### PR DESCRIPTION
**Pull request by TuomasTaipale**
Fix a few bugs related to class-of-service hash-queues:

  - In `odp_cls_cos_create()`, use the passed queue parameters as per
    the specification instead of local parameters and always set
    `queue_group` flag based on number of queues.
  - Destroy class-of-service hash-queues in `odp_cos_destroy()`.
  - When printing class-of-service information in `odp_cls_print_all()`,
    take potential hash-queues into account.

Additionally, give proper names to hash-queues created in
`odp_cls_cos_create()` to help in debugging.

v2:
- Rebased
- Fixed Matias' comments

v3:
- Rebased
- Fixed remaining comment
- Added reviewed-by tag

### CODE REVIEW

File: test/validation/api/classification/odp_classification_basic.c [hunk 1]
```diff
@@ -395,6 +395,47 @@ static void classification_test_pmr_composite_create(void)
...
+}
+
+static int check_capa_cos_hashing(void)
+{
+	odp_cls_capability_t capa;
+
+	if (odp_cls_capability(&capa) < 0)
+		return 0;
+
+	return capa.max_hash_queues > 1 && capa.hash_protocols.all_bits;
```
Comment (MatiasElo) [inline test/validation/api/classification/odp_classification_basic.c:436]: Perhaps `capa.hash_protocols.all_bits != 0` assert could be added to the test itself. It doesn't seem to make much sense to support multiple queues and zero hash protocols.

File: platform/linux-generic/odp_classification.c [hunk 2]
```diff
@@ -297,14 +297,17 @@ odp_cos_t odp_cls_cos_create(const char *name, const odp_cls_cos_param_t *param_
...
 				cos->queue_group = true;
 				cos->queue = ODP_QUEUE_INVALID;
 				_odp_cls_update_hash_proto(cos,
 							   param.hash_proto);
 				tbl_index = i * CLS_COS_QUEUE_MAX;
 				for (j = 0; j < param.num_queue; j++) {
-					queue = odp_queue_create(NULL, &cos->queue_param);
+					char name[ODP_QUEUE_NAME_LEN];
+
+					snprintf(name, sizeof(name), "odp-cos-hq-%u-%u", i, j);
```
[the remark this answers is not in the document]
Comment (TuomasTaipale) [reply platform/linux-generic/odp_classification.c]: v2.

File: platform/linux-generic/odp_classification.c [hunk 3]
```diff
@@ -1961,9 +1968,9 @@ void print_queue_ident(odp_queue_t q)
 	odp_queue_info_t info;
 
 	if (!odp_queue_info(q, &info) && strlen(info.name))
-		ODP_PRINT("%s", info.name);
+		ODP_PRINT("        %s", info.name);
 	else
-		ODP_PRINT("%" PRIx64, odp_queue_to_u64(q));
+		ODP_PRINT("        %" PRIx64, odp_queue_to_u64(q));
 
 	ODP_PRINT("\n");
```
[the remark this answers is not in the document]
Comment (TuomasTaipale) [reply platform/linux-generic/odp_classification.c]: v2.

File: test/validation/api/classification/odp_classification_basic.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (TuomasTaipale) [reply test/validation/api/classification/odp_classification_basic.c]: v2.

Comment (TuomasTaipale) [reply test/validation/api/classification/odp_classification_basic.c]: True, it's a bug really. Asserted now in the test case in v2 (fatally as passing zero in cls params might lead to weird behavior in some implementations).

**Review by MatiasElo — APPROVED** [review]
One remaining comment but otherwise looking OK.

**Pull request merged**